inc/class-config.php

```diff
@@ -0,0 +1,163 @@
+<?php
+/**
+ * Core platform config loader.
+ *
+ * @package hm-platform-core
+ */
+
+namespace HM\Platform;
+
+class Config {
+
+	/**
+	 * Primary config array.
+	 *
+	 * @var array
+	 */
+	protected static $config = [];
+
+	/**
+	 * Retrieve the configuration for HM Platform.
+	 *
+	 * The configuration is defined by merging the defaults set by modules
+	 * with any overrides present in composer.json.
+	 *
+	 * @return array Configuration data.
+	 */
+	static function get() : array {
+		if ( self::$config ) {
+			return self::$config;
+		}
+
+		self::$config = self::get_merged_defaults_and_customizations();
+
+		return self::$config;
+	}
+
+	/**
+	 * The environment type is used to override the config
+	 * as well as allow for checks within a project codebase.
+	 *
+	 * Defaults to local, this constant will be set at the stack level.
+	 *
+	 * @return string The environment type string.
+	 */
+	static function get_environment_type() : string {
+		if ( defined( 'HM_ENV_TYPE' ) ) {
+			return HM_ENV_TYPE;
+		}
+
+		return 'local';
+	}
+
+	/**
+	 * Merge the defaults and the contents of the various configuration files into a single configuration.
+	 *
+	 * @return array Configuration data.
+	 */
+	protected static function get_merged_defaults_and_customizations() : array {
+		// Default config.
+		$config = [
+			'modules' => [],
+			'environments' => [
+				'local' => [],
+				'development' => [],
+				'staging' => [],
+				'production' => [],
+			],
+		];
+
+		// @todo get registered modules & their defaults here.
+
+		if ( ! defined( 'ROOT_DIR' ) ) {
+			// phpcs:ignore
+			trigger_error( 'ROOT_DIR must be defined before loading the composer autoloader.', E_USER_WARNING );
+			return $config;
+		}
+
+		// Find composer file.
+		$composer_file = ROOT_DIR . '/composer.json';
+
+		// Look for a `composer.json` file.
+		if ( is_readable( $composer_file ) ) {
+			$config = self::get_merged_settings( $config, self::get_json_file_contents_as_array( $composer_file ) );
+		} else {
+			// phpcs:ignore
+			trigger_error( 'A composer file could not be found at ' . $composer_file, E_USER_WARNING );
+			return $config;
+		}
+
+		// Look for environment specific settings in the config and merge it in.
+		$environment = self::get_environment_type();
+		if ( isset( $config['environments'], $config['environments'][ $environment ] ) ) {
+			$config = self::get_merged_settings( $config, $config['environments'][ $environment ] );
+		}
+
+		return $config;
+	}
+
+	/**
+	 * Override settings in an existing configuration file.
+	 *
+	 * Merge customisations into a configuration file. Existing settings will be overwritten.
+	 *
+	 * @param array $config    Existing configuration.
+	 * @param array $overrides Settings to merge in.
+	 *
+	 * @return array Configuration data.
+	 */
+	protected static function get_merged_settings( array $config, array $overrides ) : array {
+		$merged = $config;
+
+		foreach ( $overrides as $key => &$value ) {
```

Comment:
Why is `$value` passed by ref here?

Reply:
I just found this on php.net for recursively merging arrays - forgot the ref is default behaviour now